crates/rspack_plugin_javascript/src/parser_and_generator/mod.rs

```diff
@@ -131,57 +116,70 @@ impl ParserAndGenerator for JavaScriptParserAndGenerator {
-        match crate::ast::parse(
-          source.to_string(),
-          syntax,
-          &resource_data.resource_path.to_string_lossy(),
-          module_type,
-        ) {
-          Ok(ast) => ast,
-          Err(e) => {
-            diagnostics.append(&mut e.into_iter().map(|e| e.boxed()).collect());
-            return result_with_diagnostics(source.to_string(), diagnostics);
-          }
-        }
-        .0
-      };
-
-    run_before_pass(&mut ast, compiler_options, &mut diagnostics)?;
-
-    let output: crate::TransformOutput = crate::ast::stringify(
-      &ast,
-      additional_data
-        .remove::<CodegenOptions>()
-        .unwrap_or_else(|| CodegenOptions::new(&module_source_map_kind, Some(true))),
-    )?;
+    let syntax = Syntax::Es(EsConfig {
+      jsx: false,
+      export_default_from: false,
+      decorators: false,
+      fn_bind: true,
+      allow_super_outside_method: true,
+      ..Default::default()
+    });
+    let source_code = if syntax.dts() {
+      // dts build result must be empty
+      RawSource::from("").boxed()
+    } else if source.source().starts_with("#!") {
+      // this is a hashbang comment
+      ConcatSource::new([RawSource::from("//").boxed(), source.clone()]).boxed()
+    } else if source.source().starts_with('\u{feff}') {
+      // bom
+      let mut s = ReplaceSource::new(source.clone());
+      s.replace(0, 3, "", None);
+      s.boxed()
+    } else {
+      source.clone()
+    };
+    let cm: Arc<swc_core::common::SourceMap> = Default::default();
+    let fm = cm.new_source_file(
+      FileName::Custom(resource_data.resource_path.to_string_lossy().to_string()),
+      source_code.source().to_string(),
+    );
+    let comments = SwcComments::default();
+    let target = ast::EsVersion::EsNext;
+    let lexer = Lexer::new(syntax, target, SourceFileInput::from(&*fm), Some(&comments));
 
-    let parse_result = match crate::ast::parse(
-      output.code.clone(),
-      syntax,
-      &resource_data.resource_path.to_string_lossy(),
+    let mut ast = match crate::ast::parse(
+      lexer.clone(),
+      &fm,
+      cm.clone(),
+      Some(comments.clone()),
       module_type,
     ) {
-      Ok(parse_result) => parse_result,
+      Ok(ast) => ast,
       Err(e) => {
         diagnostics.append(&mut e.into_iter().map(|e| e.boxed()).collect());
-        return result_with_diagnostics(output.code.clone(), diagnostics);
+        return default_with_diagnostics(source, diagnostics);
       }
     };
+    // FIXME: should not update source here
+    source = source_code;
 
-    ast = parse_result.0;
-
+    let mut semicolons = Default::default();
     ast.transform(|program, context| {
       program.visit_mut_with(&mut resolver(
         context.unresolved_mark,
         context.top_level_mark,
         false,
       ));
+      // dbg!(&resource_data.resource_path);
+      // dbg!(lexer.clone().collect_vec());
+      program.visit_with(&mut semicolon::InsertedSemicolons {
```

Comment:
Collect auto inserted semicolons. The related to whether to generate `asi_safe` code in later process. For example: 
```
import { foo } from "./foo"
(a)
foo
```
⬇️
```
import { foo } from "./foo"
(a) // asi
foo
```
⬇️
```
(a)
;((0, xxx).foo()) // generated semi
```